<a href="https://colab.research.google.com/github/spacex56/a-movie-recommendation-system/blob/master/movie_collaborative_filtering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **환경설정**

In [10]:
# Collaborative Filtering (협업 필터링 : 사용자 리뷰 기반)

In [138]:
import os
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [141]:
cd /content/drive/My Drive/github

/content/drive/My Drive/github


In [142]:


!git clone https://spacex56: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@github.com/spacex56/a-movie-recommendation-system.git

Cloning into 'a-movie-recommendation-system'...


In [143]:
cd a-movie-recommendation-system

/content/drive/MyDrive/github/a-movie-recommendation-system


In [144]:
!git config --global user.email 'alexzzang0312@gmail.com'
!git config --global user.name 'spacex56'

In [145]:
!git status

On branch master

No commits yet

nothing to commit (create/copy files and use "git add" to track)


In [151]:
!git add movie_collaborative_filtering 
!git commit -m "Connect Google Colab and Drive"

[master (root-commit) 43e3bb9] Connect Google Colab and Drive
 1 file changed, 1 insertion(+)
 create mode 100644 movie_collaborative_filtering


In [153]:
!git push -u origin master

Counting objects: 3, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (3/3), done.
Writing objects: 100% (3/3), 25.07 KiB | 3.13 MiB/s, done.
Total 3 (delta 0), reused 0 (delta 0)
To https://github.com/spacex56/a-movie-recommendation-system.git
 * [new branch]      master -> master
Branch 'master' set up to track remote branch 'master' from 'origin'.


예) 사람1본영화: up,주토피아
사람2본영화 : up,주토피아,인사이드 아웃

사람1에게 인사이드 아웃을 추천

In [11]:
pip install scikit-surprise

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [12]:
import surprise 
surprise.__version__

'1.1.1'

In [13]:
import pandas as pd
from surprise import Reader,Dataset,SVD
from surprise.model_selection import cross_validate

In [90]:
import pandas as pd
from surprise import Reader
from surprise import Dataset
from surprise.model_selection import cross_validate
from surprise import NormalPredictor
from surprise import KNNBasic
from surprise import KNNWithMeans
from surprise import KNNWithZScore
from surprise import KNNBaseline
from surprise import SVD
from surprise import BaselineOnly
from surprise import SVDpp
from surprise import NMF
from surprise import SlopeOne
from surprise import CoClustering
from surprise.accuracy import rmse
from surprise import accuracy
from surprise.model_selection import train_test_split

In [14]:
df = pd.read_csv('ratings_small.csv')
df.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [15]:
df.shape

(100004, 4)

In [16]:
df.info

<bound method DataFrame.info of         userId  movieId  rating   timestamp
0            1       31     2.5  1260759144
1            1     1029     3.0  1260759179
2            1     1061     3.0  1260759182
3            1     1129     2.0  1260759185
4            1     1172     4.0  1260759205
...        ...      ...     ...         ...
99999      671     6268     2.5  1065579370
100000     671     6269     4.0  1065149201
100001     671     6365     4.0  1070940363
100002     671     6385     2.5  1070979663
100003     671     6565     3.5  1074784724

[100004 rows x 4 columns]>

# **EDA**

Ratings Distribution 

In [17]:
from plotly.offline import init_notebook_mode, plot,iplot
import plotly.graph_objs as go
init_notebook_mode(connected=True)

data2 = df['rating'].value_counts().sort_index(ascending=False)
trace = go.Bar(x=data2.index,
               text=['{:.1f}%'.format(val) for val in (data2.values/df.shape[0]*100)],
               textposition='auto',
               textfont = dict(color='#000000'),
               y=data2.values,
               )

layout = dict(title='Distribution of {} movies-ratings'.format(df.shape[0]),
              xaxis=dict(title='Rating'),
              yaxis=dict(title='Count'))

In [18]:
data2

5.0    15095
4.5     7723
4.0    28750
3.5    10538
3.0    20064
2.5     4449
2.0     7271
1.5     1687
1.0     3326
0.5     1101
Name: rating, dtype: int64

In [19]:
fig = go.Figure(data=trace,layout=layout) 
# iplot(fig)
fig.show(renderer='colab')

x: 평가 점수 

y: 평가점수 분포 %

**Ratings Distribution by movie**

In [69]:
data = df.groupby('movieId')['rating'].count().clip(upper=50)
# movieId를 그룹화 한다 
# movieId를 그룹 마다 rating 값을 센다 (rating 값 개수세기, 더하는 것은 아님)
# clip(upper=50) -> 50은 상한값입니다. 50 이상의 값은 50으로 변경됩니다.

trace = go.Histogram(x=data.values,
                     name='Ratings',
                     xbins=dict(start=0,
                               end=50,
                               size=2))

layout = go.Layout(title='Distribution of Number of Ratings Per Movie(Clipped at 50)',
                   xaxis=dict(title='Number of Ratings Per Movie'),
                   yaxis=dict(title='Count'),
                   bargap=0.2)

fig = go.Figure(data=[trace],layout=layout)
fig.show(renderer='colab')

5개 이하의 평가를 받은 영화들이 대부분이다.

y: 영화수


x : 영화가 받은 투표수 (50개 이상 투표받은 영화는 하나로 50으로 통일)

In [47]:
df.groupby('movieId')['rating'].count().reset_index().sort_values('rating',
                                                               ascending=False)[:10]

movieId  rating
321       356     341
266       296     324
284       318     311
525       593     304
232       260     291
427       480     274
2062     2571     259
0           1     247
472       527     244
522       589     237

가장 많은 평가를 기록한 영화(Id=356)는 341개의 평가를 받았다

**Ratings Distribution By User**

In [82]:
data = df.groupby('userId')['movieId'].count().clip(upper=50)

trace = go.Histogram(x=data.values,
                     name='Ratings',
                     xbins=dict(start=0,
                                end=50,
                                size=2))

layout = go.Layout(title='Distribution of Number of ratings per user(clipped at 50)',
                   xaxis = dict(title='Ratings per user'),
                   yaxis = dict(title='Count'),
                   bargap=0.2)

fig = go.Figure(data=[trace],layout=layout)
fig.show(renderer='colab')

In [29]:
df.groupby('userId')['rating'].count().reset_index().sort_values('rating',ascending=False)[:10]

userId  rating
546     547    2391
563     564    1868
623     624    1735
14       15    1700
72       73    1610
451     452    1340
467     468    1291
379     380    1063
310     311    1019
29       30    1011

가장 많이 평가를 한 userId=547 은 2391개의 평가를 했으며, 평가 수가 저조한 user도 있다는 것을 알 수 있다. 

# **데이터 전처리 | 평가 수가 저조한 영화와 유저를 제외 시키다**

50이상은 날려야 한다라는 기준을 정한 것에 대한 근거가 부족

In [32]:
min_movies_ratings = 50
filter_movies = df['movieId'].value_counts()>min_movies_ratings
filter_movies = filter_movies[filter_movies].index.tolist()

# 50개 보다 많은 평가를 기록한 영화만을 필터링

In [42]:
len(filter_movies)

444

In [62]:
a=df['movieId'].unique()
a=a.tolist()
len(a)

9066

In [63]:
df['movieId'].nunique()

9066

In [83]:
min_user_ratings = 50
filter_users = df['userId'].value_counts() > min_user_ratings
filter_users = filter_users[filter_users].index.tolist()

# 50개 보다 많은 평가를 기록한 유저만을 필터링

In [84]:
df_new = df[(df['movieId'].isin(filter_movies)) & (df['userId'].isin(filter_users))]

print('The original data frame shape : \t{}'.format(df.shape))
print('The new data frame shape : \t{}'.format(df_new.shape))

The original data frame shape : 	(100004, 4)
The new data frame shape : 	(37915, 4)


# **Surprise 를 위한 전처리**

load_from_df() 메소드 ->  panda 데이터프레임으로부터 데이터셋을 로딩하기 위해서

Reader object 가 필요하고, rating_scale 파라메터가 특정

데이터 프레임은 사용자 id, 아이템 id, 평가에 대응하는 3개의 컬럼을 가지고 있어야 한다.

In [86]:
df_new['rating'].min()

0.5

In [87]:
df_new['rating'].max()

5.0

In [88]:
reader = Reader(rating_scale=(0.5,5)) 
#rating_scale defalt 값이 1,5 였기 때문에 scale을 조정해줌

In [89]:
data = Dataset.load_from_df(df_new[['userId','movieId','rating']],reader=reader)
# surprise 에서는 ['userId','movieId','rating'] 형식을 지원한다
data

# **모델 선택**

교차검증

정확도 척도 : RMSE 왜?

In [92]:
benchmark = []

for algorithm in [SVD(),SVDpp(),SlopeOne(),NMF(),NormalPredictor(), KNNBaseline(), KNNBasic(), KNNWithMeans(), KNNWithZScore(), BaselineOnly(), CoClustering()]:
  results = cross_validate(algorithm,data,measures=['RMSE'],cv=3,verbose=False)

  # 결과 얻기 & 알고리즘 이름을 저장
  tmp = pd.DataFrame.from_dict(results).mean(axis=0)
  tmp = tmp.append(pd.Series([str(algorithm).split(' ')[0].split('.')[-1]],index=['Algorithm']))
  benchmark.append(tmp)

pd.DataFrame(benchmark).set_index('Algorithm').sort_values('test_rmse')

Estimating biases using als...
Computing the msd similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the msd similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Estimating biases using als...


test_rmse   fit_time  test_time
Algorithm                                       
SVDpp             0.838244  28.202868   2.190120
KNNBaseline       0.841942   0.109610   1.948170
BaselineOnly      0.849851   0.099237   0.044643
KNNWithMeans      0.850475   0.062817   1.406693
SlopeOne          0.852189   0.095673   0.971791
KNNWithZScore     0.852821   0.095532   1.533077
SVD               0.853684   1.549806   0.146517
NMF               0.876140   1.452477   0.092369
CoClustering      0.890252   0.663542   0.085251
KNNBasic          0.892703   0.063709   1.283788
NormalPredictor   1.350850   0.048142   0.123974

# **SVDpp**

SVDpp 알고리즘이 가장 좋은 rmse결과를 보였다. 


따라서 SVDpp를 사용하여 훈련 및 예측을 진행하고 교대최소제곱(Alternating Least Squares,ALS) 을 사용할 것이다

In [93]:
surprise_results = pd.DataFrame(benchmark).set_index('Algorithm').sort_values('test_rmse')

In [97]:
print('Using SVDpp')
# bsl_options = {'method' : 'als'}
algo = SVDpp()
cross_validate(algo,data,measures=['RMSE'],cv=3,verbose=False)

Using SVDpp


{'test_rmse': array([0.8318401 , 0.83718503, 0.84528394]),
 'fit_time': (22.639958143234253, 22.685911417007446, 24.223272562026978),
 'test_time': (1.9165072441101074, 2.0866687297821045, 1.7266051769256592)}

In [100]:
trainset = data.build_full_trainset()

In [101]:
algo.fit(trainset)

In [105]:
df_new[df_new['userId']==2]

userId  movieId  rating  timestamp
20       2       10     4.0  835355493
21       2       17     5.0  835355681
22       2       39     5.0  835355604
23       2       47     4.0  835355552
24       2       50     4.0  835355586
25       2       52     3.0  835356031
26       2       62     3.0  835355749
27       2      110     4.0  835355532
29       2      150     5.0  835355395
30       2      153     4.0  835355441
31       2      161     3.0  835355493
32       2      165     3.0  835355441
34       2      185     3.0  835355511
36       2      208     3.0  835355511
38       2      223     1.0  835355749
39       2      225     3.0  835355552
40       2      235     3.0  835355664
42       2      253     4.0  835355511
43       2      261     4.0  835355681
44       2      265     5.0  835355697
45       2      266     5.0  835355586
48       2      292     3.0  835355492
49       2      296     4.0  835355395
50       2      300     3.0  835355532
52       2      317     2.0  835355551
54       2      339     3.0  835355492
55       2      349     4.0  835355441
56       2      350     4.0  835355697
57       2      356     3.0  835355628
58       2      357     3.0  835355749
59       2      364     3.0  835355604
60       2      367     3.0  835355619
61       2      370     2.0  835355932
64       2      377     3.0  835355710
67       2      410     3.0  835355532
68       2      454     4.0  835355604
69       2      457     3.0  835355511
71       2      474     2.0  835355828
72       2      480     4.0  835355643
73       2      485     3.0  835355918
74       2      497     3.0  835355880
75       2      500     4.0  835355731
76       2      508     4.0  835355860
77       2      509     4.0  835355719
79       2      527     4.0  835355731
81       2      539     3.0  835355767
83       2      551     5.0  835355767
86       2      586     3.0  835355790
87       2      587     3.0  835355779
88       2      588     3.0  835355441
89       2      589     5.0  835355697
90       2      590     5.0  835355395
91       2      592     5.0  835355395
92       2      593     3.0  835355511
94       2      661     4.0  835356141

In [108]:
algo.predict(2,39,5) 
# userId=1 , movieId=1029
#r_ui : 해당 영화 실제 평가 점수
# est : 해당 영화를 평가할 것으로 예측되는 점수

Prediction(uid=2, iid=39, r_ui=5, est=3.59309822569325, details={'was_impossible': False})

# **SVD**

In [109]:
svd = SVD(random_state=0)
cross_validate(svd,data,measures=['RMSE','MAE'],cv=5,verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.8503  0.8351  0.8388  0.8474  0.8556  0.8454  0.0075  
MAE (testset)     0.6472  0.6429  0.6381  0.6486  0.6515  0.6456  0.0047  
Fit time          3.11    2.24    1.76    1.65    1.47    2.05    0.59    
Test time         0.09    0.05    0.08    0.04    0.04    0.06    0.02    


{'test_rmse': array([0.85025921, 0.83506571, 0.83884461, 0.84743453, 0.8555613 ]),
 'test_mae': array([0.64717659, 0.64290827, 0.63809817, 0.64857633, 0.65146419]),
 'fit_time': (3.1129119396209717,
  2.2412142753601074,
  1.7552549839019775,
  1.6539442539215088,
  1.4733457565307617),
 'test_time': (0.08640909194946289,
  0.04795122146606445,
  0.08306455612182617,
  0.04136967658996582,
  0.03769421577453613)}

In [110]:
svd.fit(trainset)

In [111]:
svd.predict(2,39,5) 

Prediction(uid=2, iid=39, r_ui=5, est=3.7108894996582817, details={'was_impossible': False})

# **bsl**

In [112]:
print('Using ALS')
bsl_options = {'method': 'als',
               'n_epochs': 5,
               'reg_u': 12,
               'reg_i': 5
               }
bsl = BaselineOnly(bsl_options=bsl_options)
cross_validate(bsl, data, measures=['RMSE'], cv=3, verbose=False)

Using ALS
Estimating biases using als...
Estimating biases using als...
Estimating biases using als...


{'test_rmse': array([0.84835475, 0.84807487, 0.85304535]),
 'fit_time': (0.09856128692626953, 0.06379818916320801, 0.06971597671508789),
 'test_time': (0.20317363739013672, 0.04549217224121094, 0.05227947235107422)}

In [113]:
bsl.fit(trainset)

Estimating biases using als...


In [114]:
bsl.predict(2,39,5) 

Prediction(uid=2, iid=39, r_ui=5, est=3.5385224831457016, details={'was_impossible': False})

# **전처리 하지 않은 SVDpp**

In [115]:
data_old = Dataset.load_from_df(df[['userId','movieId','rating']],reader=reader)
# surprise 에서는 ['userId','movieId','rating'] 형식을 지원한다

In [119]:
trainset_old = data_old.build_full_trainset()

In [122]:
svd.fit(trainset_old)

In [126]:
bsl.fit(trainset_old)

Estimating biases using als...


In [137]:
bsl.predict(1,2294,2) 

Prediction(uid=1, iid=2294, r_ui=2, est=2.6327295252176257, details={'was_impossible': False})

In [135]:
svd.predict(1,1339,3.5)

Prediction(uid=1, iid=1339, r_ui=3.5, est=2.7070955950880933, details={'was_impossible': False})

In [130]:
df[df['userId']==1]

userId  movieId  rating   timestamp
0        1       31     2.5  1260759144
1        1     1029     3.0  1260759179
2        1     1061     3.0  1260759182
3        1     1129     2.0  1260759185
4        1     1172     4.0  1260759205
5        1     1263     2.0  1260759151
6        1     1287     2.0  1260759187
7        1     1293     2.0  1260759148
8        1     1339     3.5  1260759125
9        1     1343     2.0  1260759131
10       1     1371     2.5  1260759135
11       1     1405     1.0  1260759203
12       1     1953     4.0  1260759191
13       1     2105     4.0  1260759139
14       1     2150     3.0  1260759194
15       1     2193     2.0  1260759198
16       1     2294     2.0  1260759108
17       1     2455     2.5  1260759113
18       1     2968     1.0  1260759200
19       1     3671     3.0  1260759117